### DISCUSSION TITLE
Added Advanced Prices to intervals on /prices and /prices/current endpoints

### DISCUSSION THREAD

**Discussion by martycoxdev** [amberelectric/public-api#214] [Announcements]
Hi everyone, my name is Marty and I have recently joined Amber as a software engineer working alongside @madpilot and team.

# We have rolled out Advanced Prices!

### What are Advanced Prices?
The `AdvancedPrice` represents the Amber forecasted price prediction for a given site. These values have been "priced" for the given `siteId`, meaning that it takes the Amber forecasted wholesale price, and includes network and market fees.

### Where is this information included?
This change will add the `advancedPrice` attribute into the `CURRENT` and `FORECAST` intervals returned from the amber public API endpoints:
 * `/sites/{siteId}/prices`
 * `/sites/{siteId}/prices/current`.

Also note that it is **not** returned for the following (given these are not estimates).
 * `ACTUAL` intervals
 * `CURRENT` intervals where `estimate=false`

### Additional Information
The OpenAPI docs have been updated, and are available [here](https://app.amber.com.au/developers).

The definition of the AdvancedPrice property is as below:
```
AdvancedPrice {
      /**
       * The lower bound of our prediction band. Price includes network and market fees. (c/kWh).
       */
      low: number
      /**
       * The upper bound of our prediction band. Price includes network and market fees. (c/kWh).
       */
      high: number
}
```

An example output of the intervals returned from `prices/current` is below:
```
[
    {
        "type": "CurrentInterval",
        "date": "2024-07-05",
        "duration": 30,
        "startTime": "2024-07-05T05:30:01Z",
        "endTime": "2024-07-05T06:00:00Z",
        "nemTime": "2024-07-05T16:00:00+10:00",
        "perKwh": 28.94841,
        "renewables": 31.468,
        "spotPerKwh": 13.94436,
        "channelType": "general",
        "spikeStatus": "none",
        "tariffInformation": {
            "period": "offPeak",
            "season": "default",
            "demandWindow": false
        },
        "descriptor": "veryLow",
        "estimate": true,
        "advancedPrice": {
            "low": 29.38334,
            "high": 36.9962
        }
    },
    {
        "type": "CurrentInterval",
        "date": "2024-07-05",
        "duration": 30,
        "startTime": "2024-07-05T05:30:01Z",
        "endTime": "2024-07-05T06:00:00Z",
        "nemTime": "2024-07-05T16:00:00+10:00",
        "perKwh": -13.58623,
        "renewables": 31.468,
        "spotPerKwh": 13.94436,
        "channelType": "feedIn",
        "spikeStatus": "none",
        "descriptor": "high",
        "estimate": true,
        "advancedPrice": {
            "low": -13.98163,
            "high": -20.90241
        }
    }
]
```

**Comment by rob0101** [comment]
Thank You.

Has any analysis been done on where the typical actual price sits compared to the advanced price range ?

e.g. is it typically nearer the low, or typically the average of the two or a random distribution ?

I'm guessing the answer should vary by state too.

**Reply by madpilot** [reply]
I'll ask the data team, but generally our forecasts are more conservative, so towards the lower side.

**Comment by purcell-lab** [comment]
Great news, we have been waiting for this enhancement!

I will be injecting these values into EMHASS tonight.

**Reply by rob0101** [reply]
what value will you choose relative to the low & high ?,  or is it able to learn from both ?

I'm toying with taking the difference between the two (the range) and then use the low + 10% x range.  That'll give me something near the low.

**Reply by purcell-lab** [reply]
Nice, these are coming though on both the 5 minute and 30 minute forecasts:

```
                type        date  duration startTime   endTime   nemTime    perKwh  renewables  spotPerKwh channelType spikeStatus   tariffInformation descriptor estimate                          advancedPrice
0     ActualInterval  2024-07-05         5  08:40:01  08:45:00  18:45:00  54.87912    7.412660    29.39112     general        none  {'period': 'peak'}       high      NaN                                    NaN
1     ActualInterval  2024-07-05         5  08:45:01  08:50:00  18:50:00  55.52044    7.360969    30.00152     general        none  {'period': 'peak'}       high      NaN                                    NaN
2     ActualInterval  2024-07-05         5  08:50:01  08:55:00  18:55:00  54.39579    7.445578    28.93110     general        none  {'period': 'peak'}       high      NaN                                    NaN
3     ActualInterval  2024-07-05         5  08:55:01  09:00:00  19:00:00  52.05083    7.806241    26.69920     general        none  {'period': 'peak'}       high      NaN                                    NaN
4     ActualInterval  2024-07-05         5  09:00:01  09:05:00  19:05:00  50.77251    7.641080    25.48251     general        none  {'period': 'peak'}       high      NaN                                    NaN
5    CurrentInterval  2024-07-05         5  09:05:01  09:10:00  19:10:00  47.68905    7.772000    22.54772     general        none  {'period': 'peak'}    neutral    False                                    NaN
6   ForecastInterval  2024-07-05         5  09:10:01  09:15:00  19:15:00  47.72884    7.628000    22.58559     general        none  {'period': 'peak'}    neutral      NaN    {'low': 43.92729, 'high': 52.28821}
7   ForecastInterval  2024-07-05         5  09:15:01  09:20:00  19:20:00  47.53642    7.638000    22.40245     general        none  {'period': 'peak'}    neutral      NaN    {'low': 43.07188, 'high': 51.13757}
8   ForecastInterval  2024-07-05         5  09:20:01  09:25:00  19:25:00  46.03576    7.646000    20.97414     general        none  {'period': 'peak'}    neutral      NaN    {'low': 42.12444, 'high': 51.48597}
9   ForecastInterval  2024-07-05         5  09:25:01  09:30:00  19:30:00  46.11655    7.654000    21.05103     general        none  {'period': 'peak'}    neutral      NaN    {'low': 42.10368, 'high': 51.29542}
10  ForecastInterval  2024-07-05         5  09:30:01  09:35:00  19:35:00  47.08577    7.666000    21.97352     general        none  {'period': 'peak'}    neutral      NaN    {'low': 41.24227, 'high': 50.93933}
11  ForecastInterval  2024-07-05         5  09:35:01  09:40:00  19:40:00  46.60421    7.686000    21.51518     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.49977, 'high': 51.77723}
12  ForecastInterval  2024-07-05         5  09:40:01  09:45:00  19:45:00  45.01397    7.708000    20.00161     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.91988, 'high': 51.89819}
13  ForecastInterval  2024-07-05         5  09:45:01  09:50:00  19:50:00  45.00041    7.730000    19.98871     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.80476, 'high': 50.60242}
14  ForecastInterval  2024-07-05         5  09:50:01  09:55:00  19:55:00  44.79092    7.656000    19.78932     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.23826, 'high': 50.54798}
15  ForecastInterval  2024-07-05         5  09:55:01  10:00:00  20:00:00  39.67999    7.684000    14.92481     general        none  {'period': 'peak'}        low      NaN    {'low': 40.59631, 'high': 50.74858}
16    ActualInterval  2024-07-05         5  08:40:01  08:45:00  18:45:00 -28.07271    7.412660    29.39112      feedIn        none                 NaN       high      NaN                                    NaN
17    ActualInterval  2024-07-05         5  08:45:01  08:50:00  18:50:00 -28.65573    7.360969    30.00152      feedIn        none                 NaN       high      NaN                                    NaN
18    ActualInterval  2024-07-05         5  08:50:01  08:55:00  18:55:00 -27.63332    7.445578    28.93110      feedIn        none                 NaN       high      NaN                                    NaN
19    ActualInterval  2024-07-05         5  08:55:01  09:00:00  19:00:00 -25.50154    7.806241    26.69920      feedIn        none                 NaN       high      NaN                                    NaN
20    ActualInterval  2024-07-05         5  09:00:01  09:05:00  19:05:00 -24.33943    7.641080    25.48251      feedIn        none                 NaN       high      NaN                                    NaN
21   CurrentInterval  2024-07-05         5  09:05:01  09:10:00  19:10:00 -21.53628    7.772000    22.54772      feedIn        none                 NaN       high    False                                    NaN
22  ForecastInterval  2024-07-05         5  09:10:01  09:15:00  19:15:00 -21.57246    7.628000    22.58559      feedIn        none                 NaN       high      NaN   {'low': -18.1165, 'high': -25.71733}
23  ForecastInterval  2024-07-05         5  09:15:01  09:20:00  19:20:00 -21.39753    7.638000    22.40245      feedIn        none                 NaN       high      NaN   {'low': -17.33885, 'high': -24.6713}
24  ForecastInterval  2024-07-05         5  09:20:01  09:25:00  19:25:00 -20.03329    7.646000    20.97414      feedIn        none                 NaN       high      NaN  {'low': -16.47754, 'high': -24.98803}
25  ForecastInterval  2024-07-05         5  09:25:01  09:30:00  19:30:00 -20.10673    7.654000    21.05103      feedIn        none                 NaN       high      NaN   {'low': -16.45867, 'high': -24.8148}
26  ForecastInterval  2024-07-05         5  09:30:01  09:35:00  19:35:00 -20.98784    7.666000    21.97352      feedIn        none                 NaN       high      NaN  {'low': -15.67557, 'high': -24.49108}
27  ForecastInterval  2024-07-05         5  09:35:01  09:40:00  19:40:00 -20.55006    7.686000    21.51518      feedIn        none                 NaN       high      NaN  {'low': -15.00057, 'high': -25.25281}
28  ForecastInterval  2024-07-05         5  09:40:01  09:45:00  19:45:00 -19.10439    7.708000    20.00161      feedIn        none                 NaN       high      NaN  {'low': -15.38249, 'high': -25.36278}
29  ForecastInterval  2024-07-05         5  09:45:01  09:50:00  19:50:00 -19.09207    7.730000    19.98871      feedIn        none                 NaN       high      NaN   {'low': -15.27784, 'high': -24.1848}
30  ForecastInterval  2024-07-05         5  09:50:01  09:55:00  19:55:00 -18.90162    7.656000    19.78932      feedIn        none                 NaN       high      NaN  {'low': -14.76283, 'high': -24.13531}
31  ForecastInterval  2024-07-05         5  09:55:01  10:00:00  20:00:00 -14.25532    7.684000    14.92481      feedIn        none                 NaN       high      NaN  {'low': -15.08834, 'high': -24.31767}

               type        date  duration startTime   endTime   nemTime    perKwh  renewables  spotPerKwh channelType spikeStatus   tariffInformation descriptor estimate                          advancedPrice
0    ActualInterval  2024-07-05        30  08:30:01  09:00:00  19:00:00  54.78025    7.431000    29.29702     general        none  {'period': 'peak'}       high      NaN                                    NaN
1   CurrentInterval  2024-07-05        30  09:00:01  09:30:00  19:30:00  47.64652    7.663000    22.50724     general        none  {'period': 'peak'}    neutral     True    {'low': 44.94814, 'high': 50.77812}
2  ForecastInterval  2024-07-05        30  09:30:01  10:00:00  20:00:00  44.69588    7.688000    19.69886     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.71687, 'high': 51.08563}
3  ForecastInterval  2024-07-05        30  10:00:01  10:30:00  20:30:00  47.08628    7.926706    21.97401     general        none  {'period': 'peak'}    neutral      NaN    {'low': 40.02414, 'high': 50.23722}
4  ForecastInterval  2024-07-05        30  10:30:01  11:00:00  21:00:00  40.43866    8.236325    15.64690     general        none  {'period': 'peak'}        low      NaN     {'low': 38.9209, 'high': 48.68927}
5    ActualInterval  2024-07-05        30  08:30:01  09:00:00  19:00:00 -27.98283    7.431000    29.29702      feedIn        none                 NaN       high      NaN                                    NaN
6   CurrentInterval  2024-07-05        30  09:00:01  09:30:00  19:30:00 -21.49762    7.663000    22.50724      feedIn        none                 NaN       high     True  {'low': -19.04455, 'high': -24.34453}
7  ForecastInterval  2024-07-05        30  09:30:01  10:00:00  20:00:00 -18.81522    7.688000    19.69886      feedIn        none                 NaN       high      NaN  {'low': -15.19794, 'high': -24.62408}
8  ForecastInterval  2024-07-05        30  10:00:01  10:30:00  20:30:00 -20.98831    7.926706    21.97401      feedIn        none                 NaN       high      NaN   {'low': -14.56819, 'high': -23.8528}
9  ForecastInterval  2024-07-05        30  10:30:01  11:00:00  21:00:00 -14.94502    8.236325    15.64690      feedIn        none                 NaN       high      NaN  {'low': -13.56524, 'high': -22.44557}
```

**Reply by purcell-lab** [reply]
Have this injecting the low Advanced Price Forecast into EMHASS, let see how this performs. 

Purple line below (of course)
![Screenshot_20240705-204856](https://github.com/amberelectric/public-api/assets/79175134/80c66b99-cc8a-4203-9725-0e3f2dab01cf)

**Comment by ffrog8** [comment]
This is great news! Is there any updates required for the home assistant integration to use these advanced prices values?

**Reply by madpilot** [reply]
Yep - I need to update the code.

**Reply by ffrog8** [reply]
:) you mean that stuff doesn't just happen? 
Cheers

**Comment by rob0101** [comment]
Looking like the "low" number is pretty good for SA during a negative period.

![Screenshot 2024-07-06 at 12 29 24 PM](https://github.com/amberelectric/public-api/assets/19808721/00f95305-96f4-4306-9f9d-d4360c8b5d4e)

**Comment by rob0101** [comment]
I've been playing with different formulae for choosing a price given we now have 3: low, high, aemo.

I only have just over a day of data to work with, so things will change over time.

Right now, the best results (on SA prices) has been to choose:

a) aemo if it sits between low and high.
b) "low" if it is greater than aemo
c) average(low, high) if aemo > high.

Comparing low, high and aemo outcomes, they've roughly shared the same probability of being the nearest to the final price, so each around 33% "best" rate. However if you look at how often they are the worst choice, aemo is worst around 54% of the time,  "low" is wrong 27% of the time, "high" 19%. 

Since Amber favours "low" I've compared the outcome of replacing "low" with the formula above and it currently wins 50% of the time and is worst 27%, so a big improvement in the wins column.

None of the options predicted SA's mini spike this afternoon when solar dropped suddenly at the same time as wind was reducing.  We'd need really good weather predictions to catch those ahead of time.

**Reply by rob0101** [reply]
After 50 hours of data:

Wins:  low 14%, high 41%, aemo 45%.
Worst: low 12%, high 33%, aemo 55%

My formula is still winning out as a replacement for low, giving:   39% wins,  19% worst.

**Reply by rob0101** [reply]
Definitely seeing an advantage to choosing the aemo price if it sits between low and high.

Getting a 43% win versus 21% worst case, with a simple formula of, pick the middle value.

This can be interpreted as the aemo price being good unless it's outside the Amber forecast band.

**Comment by purcell-lab** [comment]
Also has an interesting treatment of CurrentInterval, Amber continue to publish low/high forecasts for the first 25 minutes of each interval and then remove those forecasts for the last 5 minutes as it is no longer an estimate and the full 30 minute price is locked in.

**Reply by madpilot** [reply]
Let me know if that is annoying. I’m doing the HA integration mods at the moment and I feel like it could be annoying.

Though I’m not going to add them to 5 minute current prices.

**Reply by purcell-lab** [reply]
It does complicate the automation having it disappear at minute 25, when accessing via the API.

It probably doesn't matter though for the integration, having the currentInterval in one sensor works well with or without AdvancedPrice attributes.

The forecast sensors should include all AdvancedPrice attributes.

I note the API can return the full data set from AEMO, upto 72 intervals (0330 for the next day), but the AdvancedPrice attributes are only available for the next 48 or even 49 intervals (24 hours). 

Is there a possibility to extend the AdvancedPrice forecasts to match the full AEMO intervals?

**Comment by Mijawel** [comment]
I'm a bit confused about Advanced Prices. What is the difference between 'predicted' and 'perKwh'?

{
   "type":"ForecastInterval",
   "date":"2024-09-09",
   "duration":5,
   "startTime":"2024-09-09T00:35:01Z",
   "endTime":"2024-09-09T00:40:00Z",
   "nemTime":"2024-09-09T10:40:00+10:00",
   **"perKwh":4.48901,**
   "renewables":57.335,
   "spotPerKwh":-5.02835,
   "channelType":"general",
   "spikeStatus":"none",
   "tariffInformation":{
      "period":"offPeak",
      "season":"default"
   },
   "descriptor":"extremelyLow",
   "advancedPrice":{
      "low":5.05995,
      **"predicted":5.99986,**
      "high":7.72442
   }
}

**Reply by madpilot** [reply]
`perKwh` is the current market forecast from AEMO. 

`predicted` is what Amber think the price will be.

**Reply by Mijawel** [reply]
My assumption is that Amber uses AEMO forecasts to help predict the price. 

Is the only difference that Amber combines a few models to improve the prediction?

Or that the first AEMO current market forecast doesn't include some variables (such as network tariffs, etc)

Just a bit confused because it seems like the app is using the perKwh value for the forecast. Wouldn't you guys use the predicted price if it's better?

**Reply by madpilot** [reply]
> My assumption is that Amber uses AEMO forecasts to help predict the price.
>
> Is the only difference that Amber combines a few models to improve the prediction?

Correct.

> Or that the first AEMO current market forecast doesn't include some variables (such as network tariffs, etc)

AEMO's “forecast” isn't really a forecast — it's the current state of the system. ie, if the forecasted operational demand doesn't change, then the price won't change. There are models that go into each generator's models, but the price is a second order effect.

> Just a bit confused because it seems like the app is using the perKwh value for the forecast. Wouldn't you guys use the predicted price if it's better?

It's complicated. We use it for Smartshift, but we made the decision at this point to show the AEMO prices for the moment. We'll be updating the UI at some point to show our forecasts too, but we aren't really sure what that'll look like yet.

**Reply by Mijawel** [reply]
Awesome, thanks for the detailed reply.

So essentially if I am looking for a prediction, in your opinion, the predicted price is likely to be a better choice? (I know no guarantees)

**Reply by madpilot** [reply]
Yeah, I'd rely more on the predicted attribute. It's currently a little _too_ conservative, but we are continually tuning it.

**Comment by kieranclancy** [comment]
Can you give a more precise definition of "Estimated minimum price" and "Estimated maximum price"?

Suppose `low` is 3, `predicted` is 7, and `high` is 12. Does this mean there is some probability that you expect the true value to be between `low` and `high`? For example, maybe you expect the true value will be between `low` and `high` 95% of the time? 99%? Or maybe you expect the true value to be above `low` 95% of the time, and below `high` 95% of the time? 99%?

If you look back at past predicted intervals, what percentage of them were above the `low`, what percentage of them were below the `high`, and what percentage of them were between `low` and `high`?

It would be an improvement to include this in the schema e.g.
```js
advancedPrice: {
  low: 3.00,
  aboveLowWithProb: 0.98,
  high: 12.00,
  belowHighWithProb: 0.98,
  predicted: 7.00
}
```

By the way, is `predicted` the expected value, the median prediction, or the mode (peak) of the probability distribution? I'm assuming it's the expected value? It would be good to clarify that in the schema description.
